### ISSUE TITLE
Is Fast-LIO's Kinetic Model wrong?

### ISSUE THREAD

**Issue by Huguet57** [hku-mars/FAST_LIO#113]
Hi all,
I have a worrying question about the **Kinetic Model**. I'm referring at the one referenced in the paper:

![Kinetic Model Formulation](https://user-images.githubusercontent.com/6130834/146977053-116c60cf-d1fd-4a7b-a032-2e0f8fde56d1.png)

While trying to improve on the code, I have seen that integrating the IMUs (without LiDAR) works **too** bad. Here are the results:

- Ground truth circuit (with path drawn in green):
![Ground Truth Path](https://user-images.githubusercontent.com/6130834/146978221-1dceeb3f-0af6-4df7-9fe1-eef22e6210cb.png)

- Integrating IMUs with Fast-LIO's Kinetic Model starting from the yellow X:
![Integrating IMUs with Fast-LIO's Kinetic Model](https://user-images.githubusercontent.com/6130834/146977465-d5b18777-14f5-4f5f-a988-e6d55372a36b.png)

Specially what worried me where the **curves**:
![Zoom on Curve 1](https://user-images.githubusercontent.com/6130834/146977547-ec40b72a-76ae-4565-9070-2eef23944242.png)
![Zoom on Curve 2](https://user-images.githubusercontent.com/6130834/146977558-83f40af1-fac0-49e7-8c98-9659cb78188b.png)

 What I thought is that Fast-LIO only considers the rotation to update the velocity and not the position. So if acceleration is 0 in a curve, Fast-LIO **would not update the velocity** and therefore neither the position.

Then, I tried to multiply the **rotation also on the position** (since integrating would leave the R, right?).

This is the result:

![Result multiplying the rotation on the position](https://user-images.githubusercontent.com/6130834/146978140-047f99af-2cd2-4282-9b40-eb8148373432.png)

![Ground Truth Path](https://user-images.githubusercontent.com/6130834/146978221-1dceeb3f-0af6-4df7-9fe1-eef22e6210cb.png)

I put again the Ground Truth Path. Seems way more consistent, doesn't it?

Tell me your thoughts,
Andreu

**Comment by Huguet57** [comment]
Additional screenshot of the **new proposed formulation** before closing the loop. It's really similar to the ground truth, it has the same curves.

![New formulation before closing the loop](https://user-images.githubusercontent.com/6130834/146979562-7b6d7b43-3482-4544-a900-b312995ad999.png)

With the **old formulation**:
![Old formulation before closing the loop](https://user-images.githubusercontent.com/6130834/146980096-32a76f8d-b550-4069-b8c5-c50c2515ea90.png)

**Comment by Huguet57** [comment]
Second lap with new formulation, I think this is an obvious one. This is by integrating IMUs from the beginning alone.

![Second lap with new formulation](https://user-images.githubusercontent.com/6130834/146981128-4531d6ff-8959-4830-8ca7-3bdda9f1e4c9.png)

**Comment by Huguet57** [comment]
Either that or I'm understanding wrong the meaning the value GvI.
How I am doing the integration is:

Old:
`R    *= SO3::Exp(w, dt)`
`vel  += R*((imu.a - ba) + g)*dt`
`pos += vel*dt`

New:
`R    *= SO3::Exp(w, dt)`
`vel  += R*((imu.a - ba) + g)*dt`
`pos += R*vel*dt`

**Comment by XW-HKU** [comment]
Thanks for your deep research.

I am a bit confused. The velocity in your new formular is also defined in world frame. If using the world frame velocity to update position, there should not have a rotation matrix multiplied in the left.

**Comment by Huguet57** [comment]
@XW-HKU  Yes, that's what I was thinking too this morning.
I tried rotating the velocity vector too bit by bit:

`// R ⊞ (w - bw - nw)*dt`
`Eigen::Vector3f w_hat = imu.w - this->bw;`
`this->R *= SO3::Exp(w_hat, dt);`
`// v ⊞ (R*(a - ba - na) + g)*dt`
`this->vel = SO3::Exp(w_hat, dt)*this->vel; // too fast`
`this->vel = this->R*((imu.a - this->ba) + this->g)*dt; // assuming flat surface`
`// p ⊞ v*dt`
`this->pos += this->vel*dt;`

The results are also good but it accelerates way more...

Here is the first lap:
![First lap](https://user-images.githubusercontent.com/6130834/147040526-19e1f096-2975-42e5-8df1-f86b4c67b029.png)

Here is the second lap:
![Second lap](https://user-images.githubusercontent.com/6130834/147040532-32781093-67ed-4b53-a16a-ce0a147446e5.png)

This theorically makes more sense since I'm transforming the velocity vector the same way I'm transforming the states' rotation. But the velocities are more controlled in the previous method.

**Comment by XW-HKU** [comment]
this->vel = this->R*((imu.a - this->ba) + this->g)*dt;

**Comment by Huguet57** [comment]
Oh, this is actually another separate thing. I am assuming that the car is on a flat surface so it's better to compensate for gravity directly to the IMU reading than the rotated one. I haven't thought about putting the acceleration, I will try it.

**Comment by Huguet57** [comment]
After studying hard the problem and thanks to external help and tips from my researcher I have concluded that indeed **there's nothing wrong** with Fast-LIO's model. While the "new formulation" being more aesthetically pleasing to the human eye, the below graphics talk for themselves.

The old formulation has a position error that keeps on increasing **consistently**. As expected with IMUs. The new, while seeming better to the naked eye, has an inconsistency to it that would not help when trying to model its error.

:heavy_check_mark: **Old** formulation:
- X global position error
![X position error](https://user-images.githubusercontent.com/6130834/147209066-d6fb2fd1-848c-4cda-9c64-3065e5d50033.png)
- Y global position error
![Y position error](https://user-images.githubusercontent.com/6130834/147209071-b62e315a-7018-44f5-8266-ac2a750c8e17.png)

:x: **New** formulation:

- X global position error
![X position error](https://user-images.githubusercontent.com/6130834/147209517-fc8f4295-48b0-48a4-b17a-1d04cf8b48fd.png)

- Y global position error
![Y position error](https://user-images.githubusercontent.com/6130834/147209508-d2f02bc1-385e-4a95-ae3e-b83b587240a4.png)
 
The "old formulation" is now the following (I tried use the previous values as done in a linear update):

`Eigen::Matrix3f Rtemp = R;`
`Eigen::Vector3f veltemp = vel;`
`Eigen::Vector3f postemp = pos;`

`Rtemp *= SO3::Exp(imu.w - bw, dt);`
`veltemp += (R*(imu.a - ba) - g)*dt;`
`postemp += vel*dt + 0.5*(R*(imu.a - ba) - g)*dt*dt;`

`R = Rtemp;`
`vel = veltemp;`
`pos = postemp;`

Thanks also @XW-HKU for the tips, I will keep trying to improve your amazing project for spinnings LiDARs. I am closing this.
